src/tests/quinn.rs

```diff
@@ -0,0 +1,299 @@
+                            let _x = conns.write().await.push(connection);
+                            listen_for_messages(uni_streams, peer_address, message_sender.clone());
+                        }
+                        Err(err) => {
+                            error!(
+                                "An incoming connection failed because of an error: {:?}",
+                                err
+                            );
+                        }
+                    },
+                    None => {
+                        trace!("quinn::Incoming::next() returned None. There will be no more incoming connections");
+                        break;
+                    }
+                }
+            }
+        });
+
+        Ok((
+            Peer {
+                endpoint,
+                client_cfg,
+                message_tx,
+                connections,
+            },
+            message_rx,
+        ))
+    }
+
+    fn socket_addr(&self) -> Result<SocketAddr> {
+        Ok(self.endpoint.local_addr()?)
+    }
+
+    async fn send_message(&mut self, addr: &SocketAddr, msg: Bytes) -> Result<()> {
+        let new_conn = self
+            .endpoint
+            .connect_with(self.client_cfg.clone(), addr, DOMAIN)?
+            .await?;
+        let mut stream = new_conn.connection.open_uni().await?;
+        let _ = stream.write_all(&msg).await?;
+        stream.finish().await?;
+        self.connections.write().await.push(new_conn.connection);
+        Ok(())
+    }
+
+    fn close(&self) {
+        self.endpoint.close(0_u8.into(), b"");
+    }
+}
+
+fn listen_for_messages(
+    mut uni_streams: quinn::IncomingUniStreams,
+    src: SocketAddr,
+    message_tx: ChannelSender<(SocketAddr, Bytes)>,
+) {
+    let _ = tokio::spawn(async move {
+        read_from_stream(&mut uni_streams, src, message_tx).await?;
+        trace!("The connection to {:?} has been terminated.", src);
+        Ok::<(), anyhow::Error>(())
+    });
+}
+
+async fn read_from_stream(
+    uni_streams: &mut quinn::IncomingUniStreams,
+    peer_addr: SocketAddr,
+    message_tx: ChannelSender<(SocketAddr, Bytes)>,
+) -> Result<()> {
+    while let Some(result) = uni_streams.next().await {
+        match result {
+            Err(quinn::ConnectionError::ApplicationClosed { .. }) => {
+                trace!("Connection terminated by peer {:?}.", peer_addr);
+                break;
+            }
+            Err(err) => {
+                warn!(
+                    "Failed to read incoming message on uni-stream for peer {:?} with error: {:?}",
+                    peer_addr, err
+                );
+                break;
+            }
+            Ok(recv) => {
+                let bytes = recv.read_to_end(1024).await?;
+                message_tx.send((peer_addr, Bytes::from(bytes))).await;
+            }
+        }
+    }
+    Ok(())
+}
+
+#[tokio::test(flavor = "multi_thread")]
+#[ignore = "Will run into OOM on CI"]
```

Comment:
How much mem are we using here? 

would this pass CI w/ less hashing?

Seems worrying that this would OOM on CI. (or does it just timeout?)

Reply:
We can get this passing by reducing the number of sender and/or messages. 

It OOMs, I've seen `failed to allocate 123456 bytes` on CI with this test